x/airdrop/keeper/claim_test.go

```diff
@@ -0,0 +1,450 @@
+		expectedError             string
+	}{
+		{
+			// Claimed early middway through the first day
+			// 10 rewards on each of 3 days, 30 total rewards
+			// 50% penalty, 15 distributed, 15 forfeited
+			name:                      "claim early the first day",
+			timeOffset:                time.Hour, // one hour into first window
+			penalty:                   sdk.MustNewDecFromStr("0.5"),
+			initialAllocations:        []int64{10, 10, 10},
+			initialClaimed:            100,
+			expectedClaimed:           100 + 15,
+			expectedForfeited:         15,
+			expectedUserBalanceChange: 15,
+			initialClaimType:          types.CLAIM_DAILY,
+		},
+		{
+			// Claimed early middway through the second day
+			// 10 rewards on each of last 2 days, 20 total rewards
+			// 25% penalty, 15 distributed, 5 forfeited
+			name:                      "claim the second day",
+			timeOffset:                time.Hour * 25, // one hour into second window
+			penalty:                   sdk.MustNewDecFromStr("0.25"),
+			initialAllocations:        []int64{0, 10, 10},
+			initialClaimed:            100,
+			expectedClaimed:           100 + 15,
+			expectedForfeited:         5,
+			expectedUserBalanceChange: 15,
+			initialClaimType:          types.CLAIM_DAILY,
+		},
+		{
+			// Previous daily claims causing earlier days to be 0
+			// Claimed on the third day, 30 rewards remaining
+			// 10% penalty, 27 distributed, 3 forfeited
+			name:                      "claim with previous claims",
+			timeOffset:                time.Hour * 49, // one hour into third window
+			penalty:                   sdk.MustNewDecFromStr("0.1"),
+			initialAllocations:        []int64{0, 0, 10, 20},
+			initialClaimed:            100,
+			expectedClaimed:           100 + 27,
+			expectedUserBalanceChange: 27,
+			expectedForfeited:         3,
+			initialClaimType:          types.CLAIM_DAILY,
+		},
+		{
+			// Claimer already chose claim early
+			name:               "already chose to claim early",
+			timeOffset:         time.Hour,
+			initialAllocations: []int64{},
+			initialClaimed:     100,
+			initialClaimType:   types.CLAIM_EARLY,
+			expectedError:      "user has already elected claim option",
+		},
+		{
+			// Claimer has no rewards remaining
+			name:                      "no rewards",
+			timeOffset:                time.Hour * 49, // one hour into third window
+			initialAllocations:        []int64{0, 0, 0, 0},
+			initialClaimed:            100,
+			expectedClaimed:           100,
+			expectedUserBalanceChange: 0,
+			initialClaimType:          types.CLAIM_DAILY,
+			expectedError:             "no unclaimed rewards",
+		},
+		{
+			// Claimed 1 hour before the airdrop started
+			name:               "airdrop not started yet",
+			timeOffset:         (-1 * time.Hour), // before airdrop start
+			initialAllocations: []int64{},
+			initialClaimed:     100,
+			expectedError:      "airdrop distribution has not started",
+		},
+		{
+			// Claimed well after the decision deadline
```

Comment:
Are we allowing ClaimEarly throughout the airdrop? Thought the deadline to choose that option was earlier than the end of the airdrop. This test implies the deadlines are the same.

Reply:
can you say more about why this test implies the deadlines are the same?

Reply:
Aren't there two separate deadlines? The first is the time at which you can no longer claim early, and the second is the time at which you can no longer claim. Or do we only have the latter

Reply:
Ah understood. Then we're good here. Ty for explaining